subsys/mpsl/clock_ctrl/mpsl_clock_ctrl.c

```diff
@@ -0,0 +1,317 @@
+ * In case of nrf2 clock control driver the function can time out if there is no response from
+ * remote clock controller.
+ *
+ * @note For nRF54H SoC series waiting for LFCLK can't block the system work queue. The nrf2 clock
+ *       control driver can return -TIMEDOUT due not handled response from sysctrl.
+ *
+ * @retval 0 LFCLK is ready.
+ * @retval -NRF_EINVAL There were no LFCLK request.
+ * @retval -NRF_EFAULT LFCLK request failed.
+ */
+static int32_t m_lfclk_wait(void)
+{
+	int32_t err;
+	int32_t res;
+
+	if (m_lfclk_state.m_clk_ready == true) {
+		return 0;
+	}
+
+	/* Check if lfclk has been requested */
+	if (m_lfclk_state.m_clk_refcnt <= 0) {
+		return -NRF_EINVAL;
+	}
+
+	do {
+		err = sys_notify_fetch_result(&m_lfclk_state.cli.notify, &res);
+		if (!err && res) {
+			/* Due to NCSDK-31169, temporarily allow for LFCLK request to timeout.
+			 * That doens't break anything now because the LFCLK requested clock is
+			 * 500PPM and such LFCLK should be running from boot of the radio core.
+			 */
+			if (IS_ENABLED(CLOCK_CONTROL_NRF2) && (res == -ETIMEDOUT)) {
+				LOG_WRN("LFCLK could not be started: %d", res);
+				return 0;
+			}
+
+			__ASSERT(false, "LFCLK could not be started, reason: %d", res);
+			/* Possible failure reasons:
+			 *  # -ERRTIMEDOUT - nRFS service timeout
+			 *  # -EIO - nRFS service error
+			 *  # -ENXIO - request rejected
+			 * All these mean failure for MPLS.
+			 */
+			return -NRF_EFAULT;
+		}
+
+		/* The function is executed in POST_KERNEL hence sleep is allowed */
+		k_msleep(1);
+	} while (err == -EAGAIN);
+
+	/* In case there is some other error. */
+	if (err < 0) {
+		return err;
+	}
+
+	atomic_set(&m_lfclk_state.m_clk_ready, true);
+
+	return 0;
+}
+
+#if defined(CONFIG_CLOCK_CONTROL_NRF)
+
+static void m_hfclk_request(void)
+{
+	/* The z_nrf_clock_bt_ctlr_hf_request doesn't count references to HFCLK,
+	 * it is caller responsibility handle requests and releases counting.
+	 */
+	if (atomic_inc(&m_hfclk_refcnt) > 0) {
+		return;
+	}
+
+	z_nrf_clock_bt_ctlr_hf_request();
+}
+
+static void m_hfclk_release(void)
+{
+	/* The z_nrf_clock_bt_ctlr_hf_request doesn't count references to HFCLK,
+	 * it is caller responsibility to do not release the clock if there is
+	 * other request pending.
+	 */
+	if (m_hfclk_refcnt < 1) {
+		return;
+	}
+
+	if (atomic_dec(&m_hfclk_refcnt) > 1) {
+		return;
+	}
+
+	z_nrf_clock_bt_ctlr_hf_release();
+}
+
+static bool m_hfclk_is_running(void)
+{
+	/* As of now assume the HFCLK is runnig after the request was put */
+	return ((m_hfclk_refcnt > 0) ? true : false);
+}
+
+static void m_lfclk_calibration_start(void)
+{
+	if (IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_DRIVER_CALIBRATION)) {
+		z_nrf_clock_calibration_force_start();
+	}
+}
+
+static bool m_lfclk_calibration_is_enabled(void)
+{
+	if (IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_DRIVER_CALIBRATION)) {
+		return z_nrf_clock_calibration_is_in_progress();
```

Comment:
Calibration "is enabled" is something else than "is in progress", So I would assume the correct thing to do is to change this to
```suggestion
		return true;
```

Reply:
No, there is a difference in naming convention between nrf clock control and former mpsl clock control. What is `enabled` for mpsl clock control refers to: https://github.com/nrfconnect/sdk-zephyr/blob/693769a5c7354efce42af62b6c7f68f77c6678b7/drivers/clock_control/nrf_clock_calibration.c#L23-L24.
the nrf clock control uses this global to store state: https://github.com/nrfconnect/sdk-zephyr/blob/693769a5c7354efce42af62b6c7f68f77c6678b7/drivers/clock_control/nrf_clock_calibration.c#L34 and 
the function used returns its value: https://github.com/nrfconnect/sdk-zephyr/blob/693769a5c7354efce42af62b6c7f68f77c6678b7/drivers/clock_control/nrf_clock_calibration.c#L298-L301

Reply:
I'm a bit confused here. `cal_process_in_progress ` in nrf_clock_calibration.c seems to be cleared in `start_cycle`. That is, is looks like the function only returns true when calibration is actually running.

I would expect `m_lfclk_calibration_is_enabled` to return `true` statically when calibration is enabled

Reply:
Discussed offline, the correct approach is to `return true;` if `CONFIG_CLOCK_CONTROL_NRF_DRIVER_CALIBRATION` is enabled.